precompile/precompileconfig/config.go

```diff
@@ -25,3 +31,60 @@ type Config interface {
 	// Verify is called on startup and an error is treated as fatal. Configure can assume the Config has passed verification.
 	Verify() error
 }
+
+// PrecompilePredicateContext is the context passed in to the PrecompilePredicater interface.
+type PrecompilePredicateContext struct {
+	SnowCtx *snow.Context
+}
+
+// PrecompilePredicater is an optional interface for StatefulPrecompileContracts to implement.
+// If implemented, the predicate will be enforced on every transaction in a block, prior to
+// the blcok's execution.
+// If VerifyPredicate returns an error, the block will fail verification with no further processing.
+// WARNING: If you are implementing a custom precompile, beware that subnet-evm
+// will not maintain backwards compatibility of this interface and your code should not
+// rely on this. Designed for use only by precompiles that ship with subnet-evm.
+type PrecompilePredicater interface {
+	VerifyPredicate(predicateContext *PrecompilePredicateContext, storageSlots []byte) error
+}
+
+// ProposerPredicateContext is the context passed in to the ProposerPredicater interface to verify
+// a precompile predicate within a specific ProposerVM wrapper.
+type ProposerPredicateContext struct {
+	PrecompilePredicateContext
+	// ProposerVMBlockCtx defines the ProposerVM context the predicate is verified within
+	ProposerVMBlockCtx *block.Context
+}
+
+// ProposerPredicater is an optional interface for StatefulPrecompiledContracts to implement.
+// If implemented, the predicate will be enforced on every transaction in a block, prior to
+// the block's execution.
+// If VerifyPredicate returns an error, the block will fail verification with no further processing.
+// Note: ProposerVMBlockCtx is guaranteed to be non-nil.
+// WARNING: If you are implementing a custom precompile, beware that subnet-evm
+// will not maintain backwards compatibility of this interface and your code should not
+// rely on this. Designed for use only by precompiles that ship with subnet-evm.
+type ProposerPredicater interface {
```

Comment:
I'm a bit confused about the difference between `ProposerPredicate` vs `PrecompilePredicate`. will precompiles implement both of these interfaces?

Reply:
Adding a comment for this. Precompiles should implement `ProposerPredicate` iff they depend on the ProposerVM Block Context ie. Warp.